ships/0021-local-source-upload.md

```diff
@@ -0,0 +1,348 @@
+---
+title: local-source-upload
+authors:
+  - @otaviof
+reviewers:
+  - @alicerum
+  - @adambkaplan
+  - @gabemontero
+  - @qu1queee
+approvers:
+creation-date: 2021-06-23
+last-updated: 2021-07-22
+---
+
+# Source Upload
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [docs](/docs/)
+
+## Summary
+
+Proposal to improve developer-experience by allowing to upload a local repository clone to the
+Kubernetes cluster, and use Shipwright Build Controller to produce a container-image out of it.
+
+As a developer, I can try Shipwright before opening a pull-request, and take advantage of the cluster
+resources to create the image. Effectively moving shipwright closer to the developer's inner-loop, in
+other words closer to the frequent activities while writing code ([1][InnerLoopA], [2][InnerLoopB]),
+in other words we are "moving left before _shipping-right_".
+
+## Motivation
+
+Shipwright Build-Controller allows teams to build container-images out of Git repositories, register
+and track their application builds in a Kubernetes native environment. This enhancement proposal
+extends this ability for developers to upload uncommitted changes and use the same machinery during
+the inner-loop as well.
+
+As a developer you can try out changes before opening a pull-request, make sure the final
+container-image will work out as expected, and effectively improve the developer experience.
+
+The approach proposed here is complementary to existing [EP][LocalSourceCodeEP] detailing local
+source upload to a container-registry and later on self-extracting the image before the actual
+container build process starts.
+
+### Goals
+
+- Improve developer experience, local clone can produce a container-image using Shipwright machinery
+- Move Shipwright CLI closer to developer's inner-loop
+- Complementary to [upload-to-container-registry (EP)][LocalSourceCodeEP], uploading data directly to
+the cluster instead
+
+## Proposal
+
+This proposal includes changes in Shipwright components, and adds a new helper application to aid
+data upload process.
+
+The user-experience using the CLI should be as simple as the following example, assuming the user is
+present at the same directory than the source code, and a `Build` named `nodejs-ex` exists.
+
+```bash
+shp buildrun upload nodejs-ex --to-cluster
+```
+
+Here, a new sub-command, `upload` is introduced and the user may choose `--to-cluster` or
+`--to-registry` options, to respectively, upload to the cluster directly or upload to
+container-registry.
```

Comment:
I agree with the goal that the user experience should be simple. I have to admit that I have not so much looked at our CLI in detail so far. But, the command you provide here is maybe over-simplifying:

I assume we still need a Build that specifies necessary details like strategy and output. Now, the strategy also defines what the source is and by that I guess will also define which "local source mode" is being used (particularely, the registry-based approach needs an image-url and optionally a pull secret to be functional).

Your command is for a BuildRun. At this point, the decision for streaming vs registry has already been made as the Build already exists. I therefore think that the decision between `--to-cluster` and `--to-registry` would too late here. Assuming `nodejs-ex` refers to the Build, would the CLI then maybe lookup this Build, determine that it is local source and which approach, and then send the current working directory?

Reply:
> I agree with the goal that the user experience should be simple. I have to admit that I have not so much looked at our CLI in detail so far. But, the command you provide here is maybe over-simplifying:
> 
> I assume we still need a Build that specifies necessary details like strategy and output.

For the record, the output can be overwritten in the `BuildRun`.

> Now, the strategy also defines what the source is and by that I guess will also define which "local source mode" is being used (particularely, the registry-based approach needs an image-url and optionally a pull secret to be functional).

That's part I would like to dissect. How practical is forcing the developer to create a `Build` and then a `BuildRun` to build something from local source?

In my view, the logical analogies here are: a `Build` represents the repository the repository and strategy, while the `BuildRun` is responsible to instantiate the build process and overwrite certain attributes. So, as a developer, I see the local-source as a way to create an image out of a local *clone* of a repository (`Build`), and since the local-source is "overwriting" the data it would acquire from the Git repository, this setting should be concentrated on the `BuildRun` as well.

Furthermore, would it make sense declaring a `Build` without a Git repository? If that's the case, the `Build` would only define the Strategy, since the all other settings, like `output` could be overwritten by the `BuildRun` instance.

Please consider the [demo (YT)](https://youtu.be/btgHUUMfSTo), it tackles the developer-experience, and the high-level abstractions.